Constraint with a namespaceSelector that generates VAPB returns nil

Everything went well up to the creation of the VAP, thank you!
However, as mentioned above, when I specified namespaceSelector and created the CRD and VAPB, I'm encountering a nil pointer dereference.

>But another issue is that in v3.17.1, when trying to create a VAPB with a specified namespaceSelector, it results in a nil pointer dereference. It seems that the code fix for this part has been released in v3.18.0, but when I try it, the above problem occurs.
To reproduce the issue, after applying the constraintTemplate from [the demo](https://github.com/open-policy-agent/gatekeeper/blob/master/demo/scoped-enforcement-actions/owner_must_be_provided.yaml), apply the following manifest. This just specifies the namespaceSelector meaninglessly, but in my use case, I use it to specify the namespace where the pod is running.

```yaml
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sRequiredLabels
metadata:
  name: all-must-have-owner
spec:
  enforcementAction: scoped
  scopedEnforcementActions:
  - action: deny
    enforcementPoints:
    - name: validation.gatekeeper.sh
    - name: audit.gatekeeper.sh
  - action: warn
    enforcementPoints:
    - name: vap.k8s.io
  match:
    kinds:
      - apiGroups: [""]
        kinds: ["Namespace"]
    namespaceSelector:
      matchExpressions:
        - key: "example.com"
          operator: Exists
  parameters:
    message: "All namespaces must have an `owner` label that points to your company username"
    labels:
      - key: owner
        allowedRegex: "^[a-zA-Z]+.agilebank.demo$"
```

latest logs is following

```
{"level":"info","ts":1734390069.7823236,"msg":"Starting workers","controller":"config-status-controller","worker count":1}
{"level":"info","ts":1734390069.7823648,"msg":"Starting workers","controller":"constraint-status-controller","worker count":1}
{"level":"info","ts":1734390069.7824092,"msg":"Starting workers","controller":"assign-controller","worker count":1}
{"level":"info","ts":1734390069.7845383,"msg":"Starting workers","controller":"expansion-template-controller","worker count":1}
{"level":"info","ts":1734390069.7848053,"msg":"Starting workers","controller":"mutator-status-controller","worker count":1}
{"level":"info","ts":1734390069.7848125,"msg":"Starting workers","controller":"assignimage-controller","worker count":1}
{"level":"info","ts":1734390069.7848263,"msg":"Starting workers","controller":"constrainttemplate-controller","worker count":1}
{"level":"info","ts":1734390069.7919548,"logger":"watch-manager","msg":"replaying events","registrar":"constrainttemplate-controller-status","gvk":"constraints.gatekeeper.sh/v1beta1, Kind=IngressClassRestriction"}
{"level":"info","ts":1734390069.8187132,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"IngressClassRestriction","name":"deny-unconfigured-ingressclass"}}
{"level":"info","ts":1734390069.8317344,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"IngressClassRestriction","name":"deny-unconfigured-ingressclass"}}
{"level":"info","ts":1734390069.8468244,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"IngressClassRestriction","name":"deny-unconfigured-internal-ingressclass"}}
{"level":"info","ts":1734390069.8613553,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"IngressClassRestriction","name":"deny-unconfigured-internal-ingressclass"}}
{"level":"info","ts":1734390069.8699944,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"IngressClassRestriction","name":"deny-unconfigured-internal-ingressclass"}}
{"level":"info","ts":1734390069.910832,"logger":"watch-manager","msg":"replaying events","registrar":"constrainttemplate-controller-status","gvk":"constraints.gatekeeper.sh/v1beta1, Kind=RestrictPrivilegedAndHostpath"}
{"level":"info","ts":1734390069.9420145,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"RestrictPrivilegedAndHostpath","name":"restrict-privileged-and-hostpath"}}
{"level":"info","ts":1734390069.9673939,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"RestrictPrivilegedAndHostpath","name":"restrict-privileged-and-hostpath"}}
{"level":"info","ts":1734390069.9831762,"logger":"controller","msg":"handling constraint status update","process":"constraint_status_controller","instance":{"apiVersion":"constraints.gatekeeper.sh/v1beta1","kind":"RestrictPrivilegedAndHostpath","name":"restrict-privileged-and-hostpath"}}
{"level":"info","ts":1734390070.0688033,"msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"constraint-controller","object":{"name":"gvk:RestrictPrivilegedAndHostpath.v1beta1.constraints.gatekeeper.sh:restrict-privileged-and-hostpath"},"namespace":"","name":"gvk:RestrictPrivilegedAndHostpath.v1beta1.constraints.gatekeeper.sh:restrict-privileged-and-hostpath","reconcileID":"19286d36-d592-495c-a95b-c5cfc7680404"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x21bb16d]
goroutine 208 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:111 +0x1da
panic({0x24b62c0?, 0x45694c0?})
	/usr/local/go/src/runtime/panic.go:785 +0x132
github.com/open-policy-agent/gatekeeper/v3/pkg/controller/constraint.v1beta1ToV1(0xc004f5c140)
	/go/src/github.com/open-policy-agent/gatekeeper/pkg/controller/constraint/constraint_controller.go:744 +0x36d
github.com/open-policy-agent/gatekeeper/v3/pkg/controller/constraint.getRunTimeVAPBinding(0xc000a85100?, 0xc004f5c140, {0x0?, 0x0?})
	/go/src/github.com/open-policy-agent/gatekeeper/pkg/controller/constraint/constraint_controller.go:691 +0x279
github.com/open-policy-agent/gatekeeper/v3/pkg/controller/constraint.(*ReconcileConstraint).manageVAPB(0xc0005c4000, {0x2e50d78, 0xc004f26420}, {0xc0054e9840, 0x6}, 0xc000a85100, 0xc004f10f00)
	/go/src/github.com/open-policy-agent/gatekeeper/pkg/controller/constraint/constraint_controller.go:583 +0x13e9
github.com/open-policy-agent/gatekeeper/v3/pkg/controller/constraint.(*ReconcileConstraint).Reconcile(0xc0005c4000, {0x2e50d78, 0xc004f26420}, {{{0x0, 0x0}, {0xc003b30b60, 0x64}}})
	/go/src/github.com/open-policy-agent/gatekeeper/pkg/controller/constraint/constraint_controller.go:331 +0x13a5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2e59820?, {0x2e50d78?, 0xc004f26420?}, {{{0x0?, 0xb?}, {0xc003b30b60?, 0x0?}}})
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0xa5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000d0e210, {0x2e50db0, 0xc0005d45f0}, {0x25cd4c0, 0xc00233f760})
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x39c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000d0e210, {0x2e50db0, 0xc0005d45f0})
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:261 +0x19d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:222 +0x73
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 203
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:218 +0x46c
```

I suspect that there's something wrong with [this part of the code](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/drivers/k8scel/transform/make_vap_objects.go#L125-L137).

When I specify only match.kind as in the demo, it works without any problems.
Since the issue in this thread has been resolved, should I open a separate issue?

_Originally posted by @mmmommm in https://github.com/open-policy-agent/gatekeeper/issues/3745#issuecomment-2547110393_

--- a/pkg/controller/constraint/constraint_controller.go
+++ b/pkg/controller/constraint/constraint_controller.go
@@ -737,6 +737,7 @@ func v1beta1ToV1(v1beta1Obj *admissionregistrationv1beta1.ValidatingAdmissionPol
 
 	obj.Spec.ValidationActions = actions
 	if v1beta1Obj.Spec.MatchResources != nil {
+		obj.Spec.MatchResources = &admissionregistrationv1.MatchResources{}
 		if v1beta1Obj.Spec.MatchResources.ObjectSelector != nil {
 			obj.Spec.MatchResources.ObjectSelector = v1beta1Obj.Spec.MatchResources.ObjectSelector
 		}

--- a/pkg/controller/constraint/constraint_controller_test.go
+++ b/pkg/controller/constraint/constraint_controller_test.go
@@ -3,6 +3,7 @@ package constraint
 import (
 	"context"
 	"errors"
+	"fmt"
 	"reflect"
 	"strings"
 	"testing"
@@ -14,9 +15,12 @@ import (
 	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
 	constraintstatusv1beta1 "github.com/open-policy-agent/gatekeeper/v3/apis/status/v1beta1"
 	celSchema "github.com/open-policy-agent/gatekeeper/v3/pkg/drivers/k8scel/schema"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/drivers/k8scel/transform"
 	"github.com/open-policy-agent/gatekeeper/v3/pkg/metrics"
 	"github.com/open-policy-agent/gatekeeper/v3/pkg/target"
 	"github.com/open-policy-agent/gatekeeper/v3/pkg/util"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/utils/ptr"
@@ -614,6 +618,104 @@ func TestReportErrorOnConstraintStatus(t *testing.T) {
 
 			if reflect.DeepEqual(tt.status.Status.Errors, tt.expectedStatusError) {
 				t.Errorf("expected status errors %v, got %v", tt.expectedStatusError, tt.status.Status.Errors)
+			}
+		})
+	}
+}
+
+func TestV1beta1ToV1(t *testing.T) {
+	tests := []struct {
+		name          string
+		v1beta1Obj    *admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding
+		expectedObj   *admissionregistrationv1.ValidatingAdmissionPolicyBinding
+		expectedError error
+	}{
+		{
+			name: "valid conversion",
+			v1beta1Obj: &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-binding",
+				},
+				Spec: admissionregistrationv1beta1.ValidatingAdmissionPolicyBindingSpec{
+					PolicyName: "test-policy",
+					ParamRef: &admissionregistrationv1beta1.ParamRef{
+						Name: "test-param",
+					},
+					ValidationActions: []admissionregistrationv1beta1.ValidationAction{
+						admissionregistrationv1beta1.Deny,
+						admissionregistrationv1beta1.Warn,
+					},
+					MatchResources: &admissionregistrationv1beta1.MatchResources{
+						ObjectSelector: &metav1.LabelSelector{
+							MatchLabels: map[string]string{"key": "value"},
+						},
+						NamespaceSelector: &metav1.LabelSelector{
+							MatchLabels: map[string]string{"key": "value"},
+						},
+					},
+				},
+			},
+			expectedObj: &admissionregistrationv1.ValidatingAdmissionPolicyBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-binding",
+				},
+				Spec: admissionregistrationv1.ValidatingAdmissionPolicyBindingSpec{
+					PolicyName: "test-policy",
+					ParamRef: &admissionregistrationv1.ParamRef{
+						Name:                    "test-param",
+						ParameterNotFoundAction: ptr.To[admissionregistrationv1.ParameterNotFoundActionType](admissionregistrationv1.AllowAction),
+					},
+					ValidationActions: []admissionregistrationv1.ValidationAction{
+						admissionregistrationv1.Deny,
+						admissionregistrationv1.Warn,
+					},
+					MatchResources: &admissionregistrationv1.MatchResources{
+						ObjectSelector: &metav1.LabelSelector{
+							MatchLabels: map[string]string{"key": "value"},
+						},
+						NamespaceSelector: &metav1.LabelSelector{
+							MatchLabels: map[string]string{"key": "value"},
+						},
+					},
+				},
+			},
+			expectedError: nil,
+		},
+		{
+			name: "unrecognized enforcement action",
+			v1beta1Obj: &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-binding",
+				},
+				Spec: admissionregistrationv1beta1.ValidatingAdmissionPolicyBindingSpec{
+					PolicyName: "test-policy",
+					ParamRef: &admissionregistrationv1beta1.ParamRef{
+						Name: "test-param",
+					},
+					ValidationActions: []admissionregistrationv1beta1.ValidationAction{
+						"unknown",
+					},
+				},
+			},
+			expectedObj:   nil,
+			expectedError: fmt.Errorf("%w: unrecognized enforcement action unknown, must be `warn` or `deny`", transform.ErrBadEnforcementAction),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			obj, err := v1beta1ToV1(tt.v1beta1Obj)
+			if err != nil && tt.expectedError == nil {
+				t.Fatalf("expected no error, got %v", err)
+			}
+			if err == nil && tt.expectedError != nil {
+				t.Fatalf("expected error %v, got none", tt.expectedError)
+			}
+			if err != nil && tt.expectedError != nil && err.Error() != tt.expectedError.Error() {
+				t.Fatalf("expected error %v, got %v", tt.expectedError, err)
+			}
+			if !reflect.DeepEqual(obj, tt.expectedObj) {
+				t.Errorf("expected object %v, got %v", tt.expectedObj, obj)
 			}
 		})
 	}

